src/postprocess/phonon.jl

```diff
@@ -5,26 +5,89 @@ function dynmat_red_to_cart(model::Model, dynamical_matrix)
     # The dynamical matrix `D` acts on vectors `dr` and gives a covector `dF`:
     #   dF = D · dr.
     # Thus the transformation between reduced and Cartesian coordinates is not a comatrix.
-    # To transform `dynamical_matrix` from reduced coordinates to `cart_mat` in Cartesian
+    # To transform `dynamical_matrix` from reduced coordinates to `dynmat_cart` in Cartesian
     # coordinates, we write
     #   dr_cart = lattice · dr_red,
     #   ⇒ dr_redᵀ · D_red · dr_red = dr_cartᵀ · lattice⁻ᵀ · D_red · lattice⁻¹ · dr_cart
     #                              = dr_cartᵀ · D_cart · dr_cart
     #   ⇒ D_cart = lattice⁻ᵀ · D_red · lattice⁻¹.
 
-    cart_mat = zero.(dynamical_matrix)
-    for τ = 1:size(cart_mat, 2), η = 1:size(cart_mat, 4)
-        cart_mat[:, η, :, τ] = inv_lattice' * dynamical_matrix[:, η, :, τ] * inv_lattice
+    dynmat_cart = zero.(dynamical_matrix)
+    for s = 1:size(dynmat_cart, 2), α = 1:size(dynmat_cart, 4)
+        dynmat_cart[:, α, :, s] = inv_lattice' * dynamical_matrix[:, α, :, s] * inv_lattice
     end
-    cart_mat
+    dynmat_cart
 end
 
+# Create a ``3×n_{\rm atoms}×3×n_{\rm atoms}`` tensor (for consistency with the format of
+# dynamical matrices) equivalent to a diagonal matrix with the atomic masses of the atoms on
+# the diagonal.
+function get_mass_matrix(basis::PlaneWaveBasis{T}) where {T}
+    model = basis.model
+    positions = model.positions
+    n_atoms = length(positions)
+
+    atoms_mass = atomic_mass.(model.atoms)
+    any(iszero.(atoms_mass)) && @error "Some elements have unknown masses"
+    masses = zeros(T, 3, n_atoms, 3, n_atoms)
+    for s in eachindex(atoms_mass)
+        masses[:, s, :, s] = atoms_mass[s] * I(3)
+    end
+    masses
+end
+
+@doc raw"""
+Solve the eigenproblem for a dynamical matrix: returns the `frequencies` and eigenvectors
+(`vectors`).
 """
+function phonon_modes(basis::PlaneWaveBasis{T}, dynamical_matrix) where {T}
+    n_atoms = length(basis.model.positions)
+    mass_matrix = reshape(get_mass_matrix(basis), 3*n_atoms, 3*n_atoms)
+
+    res = eigen(reshape(dynamical_matrix, 3*n_atoms, 3*n_atoms), mass_matrix)
+    norm(imag(res.values)) > sqrt(eps(T)) &&
+        @warn "Some eigenvalues of the dynmaical matrix have a large imaginary part"
+
+    signs = sign.(real(res.values))
+    frequencies = signs .* sqrt.(abs.(real(res.values)))
+
+    (; frequencies, res.vectors)
+end
+function phonon_modes_cart(basis::PlaneWaveBasis{T}, dynamical_matrix) where {T}
+    dynmat_cart = dynmat_red_to_cart(basis.model, dynamical_matrix)
+    phonon_modes(basis, dynmat_cart)
+end
+
+@doc raw"""
 Compute the dynamical matrix in the form of a ``3×n_{\rm atoms}×3×n_{\rm atoms}`` tensor
 in reduced coordinates.
 """
-@timing function compute_dynmat(basis::PlaneWaveBasis, ψ, occupation; kwargs...)
-    dynmats_per_term = [compute_dynmat(term, basis, ψ, occupation; kwargs...)
+@timing function compute_dynmat(basis::PlaneWaveBasis{T}, ψ, occupation; q=zero(Vec3{T}),
```

Comment:
If we call this `compute_dynmat` than we should also call the input variable to `phonon_modes` `dynmat`